### PR TITLE
fix(2050): Implement create subcommand for config

### PR DESCRIPTION
**Pull request by kkisic**
## Context
We are reconsidering sd-local config scheme.
This PR makes sd-local enable to use the new config scheme.

## Objective
- Implement `config create` subcommand to create the new config
- Refactor `config` package and `cmd/config` package
- Remove `--local` option in `cmd/config`

## References
https://github.com/screwdriver-cd/screwdriver/issues/2050

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: config/config_test.go [hunk 1]
```diff
@@ -71,54 +72,229 @@ func TestCreateConfig(t *testing.T) {
...
+		assert.Equal(t, "config `default` already exists", err.Error())
 	})
 }
 
-func TestSetConfig(t *testing.T) {
-	rand.Seed(time.Now().UnixNano())
-	cnfPath := filepath.Join(testDir, ".sdlocal", fmt.Sprintf("%vconfig", rand.Int()))
-	defer os.RemoveAll(filepath.Join(testDir, ".sdlocal"))
+func TestConfigListSave(t *testing.T) {
+	t.Run("success", func(t *testing.T) {
```
Comment (kkisic) [inline config/config_test.go:248]: Is there a good way to test a failing case?

File: cmd/build.go [hunk 2]
```diff
@@ -25,7 +25,7 @@ const (
 )
 
 var (
-	configNew      = config.New
+	newConfigList  = config.NewConfigList
```
Comment (yuichi10) [inline cmd/build.go:28]: I think just `New` is better.  
Now we feel below config, but if there are other settings at the root place, ConfigList sound not good. 
```
type Config struct {
	APIURL   string   `yaml:"api-url"`
	StoreURL string   `yaml:"store-url"`
	Token    string   `yaml:"token"`
	Launcher Launcher `yaml:"launcher"`
	filePath string   `yaml:"-"`
}
```

File: cmd/build.go [hunk 3]
```diff
@@ -153,7 +153,12 @@ func newBuildCmd() *cobra.Command {
 				srcPath = scm.LocalPath()
 			}
 
-			config, err := configNew(filepath.Join(sdlocalDir, "config"))
+			configList, err := newConfigList(filepath.Join(sdlocalDir, "config"))
+			if err != nil {
+				return err
+			}
+
+			config, err := configList.Get(configList.Current)
```
Comment (yuichi10) [inline cmd/build.go:161]: I feel it is better to configList have `Current` function.

File: cmd/build.go (hunk 2 above)
Comment (kkisic) [reply cmd/build.go]: Thx.
I renamed `config.NewConfigList` to `config.New`, `Config` to `Entry`, and `ConfigList` to `Config`.

File: cmd/build.go (hunk 3 above)
Comment (kkisic) [reply cmd/build.go]: I think `Config.Current` method conflicts with `Config.Current` field.

Comment (yuichi10) [reply cmd/build.go]: I see. How about to create `Config.CurrentEntry`?

File: cmd/config/create.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+		Args: cobra.ExactArgs(1),
+		Run: func(cmd *cobra.Command, args []string) {
+			name := args[0]
+
+			path, err := filePath()
+			if err != nil {
+				logrus.Fatal(err)
+			}
+
+			configList, err := configNew(path)
```
Comment (yuichi10) [inline cmd/config/create.go:28]: is it better to use config instead of configList?

File: cmd/config/create.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+			if err != nil {
+				logrus.Fatal(err)
+			}
+
+			err = configList.Add(name)
```
Comment (yuichi10) [inline cmd/config/create.go:33]: How about `AddEntry`?

File: cmd/config/create_test.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+	cnew := configNew
+	defer func() {
+		configNew = cnew
+	}()
+	configNew = func(configPath string) (c config.Config, err error) {
+		return config.New(cnfPath)
+	}
+
+	testCase := []struct {
```
Comment (yuichi10) [inline cmd/config/create_test.go:29]: How about to add test case which try to create entry that was already created.

File: config/config.go [hunk 7]
```diff
@@ -65,48 +60,74 @@ func create(configPath string) error {
...
+		return fmt.Errorf("config `%s` already exists", name)
 	}
-	defer file.Close()
 
-	var c = configList{}
+	c.Entries[name] = newEntry()
+	return nil
+}
+
+func (c *Config) Get(name string) (*Entry, error) {
```
Comment (yuichi10) [inline config/config.go:105]: How about `Entry`.

File: cmd/config/create.go (hunk 4 above)
Comment (kkisic) [reply cmd/config/create.go]: Thx. I fixed it.

File: cmd/root_test.go [hunk 8]
```diff
@@ -36,7 +36,19 @@ func (mock mockLaunch) Kill(os.Signal) {}
 func (mock mockLaunch) Clean() {}
 
 func setup() {
-	configNew = func(confPath string) (config.Config, error) { return config.Config{}, nil }
+	configNew = func(confPath string) (config.Config, error) {
+		return config.Config{
```
Comment (yuichi10) [inline cmd/root_test.go:40]: The config.Config have deep nesting.  I think it is better to declare config.Config  at return above.

File: cmd/build.go (hunk 3 above)
Comment (kkisic) [reply cmd/build.go]: What motivates you to create `Config.CurrentEntry` function?
Are there anything to make it merit?

File: config/config.go [hunk 9]
```diff
@@ -65,48 +60,74 @@ func create(configPath string) error {
...
+}
+
 func New(configPath string) (Config, error) {
 	err := create(configPath)
 	if err != nil {
 		return Config{}, err
 	}
 
-	configList, err := newConfigList(configPath)
+	file, err := os.Open(configPath)
```
Comment (yuichi10) [inline config/config.go:78]: Don't you need to close this file?

File: config/config.go [hunk 10]
```diff
@@ -65,48 +60,74 @@ func create(configPath string) error {
...
 	}
-	defer file.Close()
 
-	var c = configList{}
+	c.Entries[name] = newEntry()
+	return nil
+}
+
+func (c *Config) Get(name string) (*Entry, error) {
+	currentEntry, exists := c.Entries[name]
```
Comment (yuichi10) [inline config/config.go:106]: Is it better to use just `entry` not currentEntry?

File: config/config.go [hunk 11]
```diff
@@ -65,48 +60,74 @@ func create(configPath string) error {
...
+	if !exists {
+		return &Entry{}, fmt.Errorf("config `%s` does not exist", name)
+	}
 
-	err = yaml.NewDecoder(file).Decode(&c)
+	return currentEntry, nil
+}
 
+func (c *Config) Save() error {
+	file, err := os.OpenFile(c.filePath, os.O_RDWR|os.O_TRUNC, 0666)
```
Comment (yuichi10) [inline config/config.go:115]: Don't you need to close this file?

File: config/config.go [hunk 12]
```diff
@@ -65,48 +60,74 @@ func create(configPath string) error {
...
+	if err != nil {
+		return err
+	}
+
+	return nil
 }
 
 // Set preserve sd-local config with new value.
-func (c *Config) Set(key, value string) error {
+func (c *Entry) Set(key, value string) error {
```
Comment (yuichi10) [inline config/config.go]: I think `e` is a good receiver name.

Comment (kkisic) [reply config/config.go]: I fixed that in https://github.com/screwdriver-cd/sd-local/pull/31/commits/44e2d54edbe88c5288bc934ae231ff428e2e4a31.

File: config/config.go (hunk 10 above)
Comment (kkisic) [reply config/config.go]: I fixed it.

File: config/config.go (hunk 9 above)
Comment (kkisic) [reply config/config.go:78]: Thx.
I fixed it.

File: cmd/root_test.go (hunk 8 above)
Comment (kkisic) [reply cmd/root_test.go:47]: I made an effort to make this nest shallow.

File: config/config.go (hunk 7 above)
Comment (kkisic) [reply config/config.go]: I fixed it.

File: cmd/config/create.go (hunk 5 above)
Comment (kkisic) [reply cmd/config/create.go]: I fixed it.

File: config/config_test.go [hunk 13]
```diff
@@ -71,58 +72,233 @@ func TestCreateConfig(t *testing.T) {
...
+		rand.Seed(time.Now().UnixNano())
+		cnfPath := filepath.Join(testDir, ".sdlocal", fmt.Sprintf("%vconfig", rand.Int()))
+		err := os.MkdirAll(filepath.Join(testDir, ".sdlocal"), 0777)
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer os.RemoveAll(filepath.Join(testDir, ".sdlocal"))
+
+		file, err := os.Create(cnfPath)
```
Comment (yuichi10) [inline config/config_test.go:257]: Don't we need to add the test case that Save() create the config file, if there is no config file?

File: config/config_test.go (hunk 1 above)
Comment (yuichi10) [reply config/config_test.go:248]: It looks tough to do failure case tests of Save function.
I think we do not need the failure case.

File: config/config_test.go [hunk 14]
```diff
@@ -198,22 +371,23 @@ func TestSetConfig(t *testing.T) {
...
-				filePath: cnfPath,
 			},
 		},
 	}
 
 	for _, tt := range testCases {
 		t.Run(tt.name, func(t *testing.T) {
-			c, _ := New(cnfPath)
+
+			c := &Entry{}
```
Comment (yuichi10) [inline config/config_test.go:389]: Is it better to use `e` instead of `c`?

File: cmd/config/create_test.go (hunk 6 above)
Comment (kkisic) [reply cmd/config/create_test.go:29]: I added that test case.

File: config/config_test.go (hunk 13 above)
Comment (kkisic) [reply config/config_test.go:257]: I think when the Config object was created, it is guaranteed that the config file exists.
So, we don't use `os.O_CREATE` in `os.OpenFile()` function of `Config.Save()` function.

File: config/config_test.go (hunk 14 above)
Comment (kkisic) [reply config/config_test.go:389]: Thx.
I fixed it.

File: cmd/build.go (hunk 3 above)
Comment (yuichi10) [reply cmd/build.go]: It's just that I felt happy to take the current settings directly.

File: config/config_test.go (hunk 13 above)
Comment (yuichi10) [reply config/config_test.go:257]: I see config.New guranteed to create config file.
https://github.com/screwdriver-cd/sd-local/blob/master/config/config.go#L26

Thanks!

File: config/config_test.go (hunk 1 above)
Comment (kkisic) [reply config/config_test.go:248]: Thx.
I will not make it.

File: config/config_test.go (hunk 14 above)
Comment (yuichi10) [reply config/config_test.go:389]: Did you fixed?

Comment (kkisic) [reply config/config_test.go]: I'm sorry, I forgot to push 🙇‍♂️

**Pull request merged**